### PR TITLE
`Development`: Enable static code analysis in local continuous integration for Java

### PR DESCRIPTION
**Pull request by mateusmm01**
#### General

- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I documented the Java code using JavaDoc style.

### Motivation and Context

At the moment, the LocalCI + LocalVC setup doesn't support Static Code Analysis, which is an important part of the automatic assessment of programming exercises in Artemis.

### Description

Static Code Analysis was enabled. The SCA configuration files for Java were already given in the `resources` directory. I had to add the path where to find the result files in `LocalCIBuildJobExecutionService` and also add the commands to the script for Gradle and Maven. Lastly, by adapting and refactoring the code from https://github.com/ls1intum/static-code-analysis-parser, we can now parse the result files for various languages and give the user feedback for their submissions.
If we enable SCA for more programming languages in the future, we will already have the corresponding parsers, too.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise with Static Code Analysis enabled

Create a Java Exercise with static code analysis enabled. Add some code that should trigger some warnings, for example, an uninitialised variable. Submit the code and check if you get a SCA warning.

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

image

### CODE REVIEW

**Review by rriyaldhi — APPROVED** [review]
Got No corresponding result available in my case, but demonstrated during the testing session and worked as expected.

**Review by vinceclifford — DISMISSED** [review]
Demonstrated during testing session, works as expected. However, sometimes I get the message "No corresponding results available"

**Review by lennart-keller — APPROVED** [review]
Demonstrated during testing session

**Review by valentin-boehm — APPROVED** [review]
Demonstrated in Testing Session

**Comment by Ceiridge** [comment]
Tested on ts6, warnings appear:
image

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
We already have the library https://github.com/ls1intum/static-code-analysis-parser. Please don't duplicate the parsers into Artemis, but use the parser library instead. If adaptations and refactorings are necessary, please open a PR over there.

**Comment by mateusmm01** [comment]
> We already have the library https://github.com/ls1intum/static-code-analysis-parser. Please don't duplicate the parsers into Artemis, but use the parser library instead. If adaptations and refactorings are necessary, please open a PR over there.

I had talked to Stephan, and he said that I should copy the code as the deployment with Maven Central is a hassle and will be unnecessary soon when we stop supporting bamboo and perhaps even Jenkins.

**Review by LenaKahle — COMMENTED** [review]
I tried to test it on legacy ts2 but it doesn't complain about anything. Is there something besides online editing + static code analysis that i need to activate?

**Comment by mateusmm01** [comment]
> I tried to test it on legacy ts2 but it doesn't complain about anything. Is there something besides online editing + static code analysis that i need to activate?

TS2 doesn‘t have the LocalCI and LocalVC system connected. Check https://confluence.ase.in.tum.de/pages/viewpage.action?pageId=25252245. I think only ts3, ts6 and ts11

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java [hunk 1]
```diff
@@ -406,18 +407,25 @@ private void scriptForJavaKotlin(ProgrammingExercise programmingExercise, String
...
                 buildScript.append("""
-                        chmod +x gradlew
-                        ./gradlew clean structuralTests
-                        if [ $? -eq 0 ]; then
-                            ./gradlew behaviorTests
-                        fi
+                        mvn clean test
+                        """);
+            }
+            if (hasStaticCodeAnalysis) {
```
Comment (laurenzfb) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java:413]:
```suggestion
           else if (hasStaticCodeAnalysis) {
```
I think sequential test runs and SCA are not supported together (add least for bamboo). You can't select both in the client anyways. But it's not blocking for me, as this script generation will be removed soon anyways

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java [hunk 2]
```diff
@@ -426,6 +434,11 @@ private void scriptForJavaKotlin(ProgrammingExercise programmingExercise, String
                         ./gradlew clean test
                         """);
             }
+            if (hasStaticCodeAnalysis) {
```
Comment (laurenzfb) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java:437]:
```suggestion
          else if (hasStaticCodeAnalysis) {
```
Same here

**Review by laurenzfb — CHANGES_REQUESTED** [review]
Tested on TS3, worked for gradle and maven exercises. Code in general LGTM. Left 2 comments. Concerning the failing bamboo server tests because of code coverage, I think you might have added some files that aren't used for Java SCA. You should probably remove them and add them in another PR that also implements the support for SCA for other languages

**Review by georgyia — APPROVED** [review]
tested manually on ts2, works as expected

**Review by AntonGeTUM — COMMENTED** [review]
Tested on TS3, works with the given examples.
However, uninitialized array variables don't trigger any warnings. Is that currently intended behavior?

**Comment by mateusmm01** [comment]
> Tested on TS3, works with the given examples. However, uninitialized array variables don't trigger any warnings. Is that currently intended behavior?

I just tested it, and it should warn you if you try using that array. Else it just tells you to avoid using unused variables.

File: src/test/java/de/tum/in/www1/artemis/staticcodeanalysis/IntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import org.xml.sax.SAXParseException;
+
+import de.tum.in.www1.artemis.service.connectors.localci.scaparser.ReportParser;
+import de.tum.in.www1.artemis.service.connectors.localci.scaparser.exception.ParserException;
+import de.tum.in.www1.artemis.service.connectors.localci.scaparser.utils.XmlUtils;
+
+/**
+ * Tests each parser with an example file
+ */
+class IntegrationTest {
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/staticcodeanalysis/IntegrationTest.java:24]: I would rename this test

Comment (reschandreas) [reply src/test/java/de/tum/in/www1/artemis/staticcodeanalysis/IntegrationTest.java]: I renamed it to `StaticCodeAnalysisIntegrationTest`, is that better?

**Comment by laurenzfb** [comment]
Added a missing test that should cover the `ParserException` class. Retested locally and everthing still works. Tests added by you also look good to me.

**Review by laurenzfb — APPROVED** [review]

**Review by reschandreas — APPROVED** [review]
demonstrated in testings session, works as expected...overall code lgtm

**Pull request merged**